vault/logical_system_activity_write_testonly.go

```diff
@@ -69,9 +75,81 @@ type multipleMonthsActivityClients struct {
 	months []*singleMonthActivityClients
 }
 
+func (s *singleMonthActivityClients) addEntityRecord(record *activity.EntityRecord, segmentIndex *int) {
+	s.clients = append(s.clients, record)
+	if segmentIndex != nil {
+		index := len(s.clients) - 1
+		s.predefinedSegments[*segmentIndex] = append(s.predefinedSegments[*segmentIndex], index)
+
+	}
+}
+
+// populateSegments converts a month of clients into a segmented map. The map's
+// keys are the segment index, and the value are the clients that were seen in
+// that index. If the value is an empty slice, then it's an empty index. If the
+// value is nil, then it's a skipped index
+func (s *singleMonthActivityClients) populateSegments() map[int][]*activity.EntityRecord {
+	segments := make(map[int][]*activity.EntityRecord)
+	ignoreIndexes := make(map[int]struct{})
+	for _, i := range s.generationParameters.SkipSegmentIndexes {
+		segments[int(i)] = nil
+		ignoreIndexes[int(i)] = struct{}{}
+	}
+	for _, i := range s.generationParameters.EmptySegmentIndexes {
+		segments[int(i)] = make([]*activity.EntityRecord, 0, 0)
+		ignoreIndexes[int(i)] = struct{}{}
+	}
+
+	// if we have predefined segments, then we can construct the map using those
+	if len(s.predefinedSegments) > 0 {
+		for segment, clientIndexes := range s.predefinedSegments {
+			clientsInSegment := make([]*activity.EntityRecord, 0, len(clientIndexes))
+			for _, idx := range clientIndexes {
+				clientsInSegment = append(clientsInSegment, s.clients[idx])
+			}
+			segments[segment] = clientsInSegment
+		}
+		return segments
+
+	}
+
+	// otherwise, we should make segments of approximately equal size
+	totalSegmentCount := 1
+	if s.generationParameters.GetNumSegments() > 0 {
+		totalSegmentCount = int(s.generationParameters.GetNumSegments())
+	}
+	usableSegmentCount := totalSegmentCount - len(s.generationParameters.SkipSegmentIndexes) - len(s.generationParameters.EmptySegmentIndexes)
```

Comment:
Nit: for brevity, it might be helpful to set these earlier in the method.
```
skipIndexes := s.generationParameters.SkipSegmentIndexes
emptyIndexes := s.generationParameters.EmptySegmentIndexes
```
and use the shorter vars throughout.